lgc/patch/NggPrimShader.cpp

```diff
@@ -1018,79 +973,93 @@ void NggPrimShader::buildPassthroughPrimShader(Function *entryPoint) {
       else
         vertexIndex = createUBfe(m_nggInputs.primData, 20, 9);
       writePerThreadDataToLds(primitiveId, vertexIndex, PrimShaderLdsRegion::DistributedPrimitiveId);
-
-      BranchInst::Create(endWritePrimIdBlock, writePrimIdBlock);
     }
 
-    // Construct ".endWritePrimId" block
-    {
-      m_builder.SetInsertPoint(endWritePrimIdBlock);
+    m_builder.CreateBr(endDistribPrimitiveIdBlock);
+  }
 
-      createFenceAndBarrier();
+  // Construct ".endDistribPrimitiveId" block
+  Value *validVertex = nullptr;
+  {
+    m_builder.SetInsertPoint(endDistribPrimitiveIdBlock);
 
-      auto vertValid = m_builder.CreateICmpULT(m_nggInputs.threadIdInWave, m_nggInputs.vertCountInWave);
-      m_builder.CreateCondBr(vertValid, readPrimIdBlock, endReadPrimIdBlock);
-    }
+    if (distributePrimitiveId)
+      createFenceAndBarrier();
 
-    // Construct ".readPrimId" block
-    Value *primitiveId = nullptr;
-    {
-      m_builder.SetInsertPoint(readPrimIdBlock);
+    validVertex = m_builder.CreateICmpULT(m_nggInputs.threadIdInSubgroup, m_nggInputs.vertCountInSubgroup);
+    m_builder.CreateCondBr(validVertex, readPrimitiveIdBlock, endReadPrimitiveIdBlock);
+  }
 
-      primitiveId = readPerThreadDataFromLds(m_builder.getInt32Ty(), m_nggInputs.threadIdInSubgroup,
-                                             PrimShaderLdsRegion::DistributedPrimitiveId);
+  // Construct ".readPrimitiveId" block
+  Value *distributedPrimitiveId = nullptr;
+  {
+    m_builder.SetInsertPoint(readPrimitiveIdBlock);
 
-      m_builder.CreateBr(endReadPrimIdBlock);
+    if (distributePrimitiveId) {
+      distributedPrimitiveId = readPerThreadDataFromLds(m_builder.getInt32Ty(), m_nggInputs.threadIdInSubgroup,
+                                                        PrimShaderLdsRegion::DistributedPrimitiveId);
     }
 
-    // Construct ".endReadPrimId" block
-    {
-      m_builder.SetInsertPoint(endReadPrimIdBlock);
+    m_builder.CreateBr(endReadPrimitiveIdBlock);
+  }
```

Comment:
I believe this code could actually be further simplified by grouping everything for the `distributePrimitiveId` case together. Basically:
```
if (distributePrimitiveId) {
  create conditional branch from entryBlock to distribPrimitiveIdBlock
  fill in distribPrimitiveIdBlock, endDistribPrimitiveIdBlock, readPrimitiveIdBlock
  create phi in endReadPrimitiveIdBlock
} else {
  create direct branch from entryBlock to endReadPrimitiveIdBlock
}
```
You could still create the basic blocks unconditionally of course, they just end up being unused. (Hmm, though you'd have to insert an `unreachable` terminator, otherwise it's invalid IR.)

Reply:
I update the change. The use of `unreachable` is very interesting :)